ZConfig/datatypes.py

```diff
@@ -277,13 +280,10 @@ def _parse_address(self, s):
 
 
 def float_conversion(v):
-    try:
-        is_str = isinstance(v, basestring)
-    except NameError:
-        # Python 3 support.
-        is_str = isinstance(v, str)
+    # float() will accept both bytes and unicode on both 2 and 3
+    is_str = isinstance(v, text_type) or isinstance(v, binary_type)
     if is_str:
-        if v.lower() in ["inf", "-inf", "nan"]:
+        if v.lower() in (u"inf", u"-inf", u"nan", b'inf', b'-inf', b'nan'):
             raise ValueError(repr(v) + " is not a portable float representation")
```

Comment:
I'm not intimately familiar with the history there, but the linked docs still contain the phrase "When passing in a string, values for NaN and Infinity may be returned, *depending on the underlying C library*" (emphasis mine). It does say that float *represents* things a certain way, but it's not clear to me that it commits to *parsing* things a certain way.

I'd be happy to see this simplified, and that could arguably be a part of the changes for this PR. It's just not entirely clear to me its safe to do so.